grpc/src/client/channel.rs

```diff
@@ -228,6 +207,47 @@ impl PersistentChannel {
             runtime,
         }
     }
+
+    /// Returns the current state of the channel. If there is no underlying active channel,
+    /// returns Idle. If `connect` is true, will create a new active channel iff none exists.
+    fn state(
+        &self,
+        connect: bool,
+    ) -> Result<ConnectivityState, Box<dyn std::error::Error + Sync + Send>> {
+        let ac = self.get_active_channel(connect)?;
+        if let Some(s) = ac.connectivity_state.cur() {
+            return Ok(s);
+        }
+        return Ok(ConnectivityState::Idle);
+    }
+
+    /// Gets the underlying active channel. If `connect` is true, will create a new channel iff
+    /// there is no active channel.
+    fn get_active_channel(
+        &self,
+        connect: bool,
+    ) -> Result<Arc<ActiveChannel>, Box<dyn std::error::Error + Sync + Send>> {
+        let mut s = self
+            .active_channel
+            .lock()
+            .map_err(|_| "Could not get channel lock.".to_string())?;
+
+        if s.is_none() {
+            if connect {
+                *s = Some(ActiveChannel::new(
+                    self.target.clone(),
+                    &self.options,
+                    self.runtime.clone(),
+                ));
+            } else {
+                return Err("No active channel.".into());
```

Comment:
This function can probably return just an Option?

Also maybe we shouldn't combine these two functionalities into one function.  They're a little disjoint.

- Anyone asking for the active channel itself will _always_ want connect=true.
- Only when requesting the current state, exiting idle mode is optional.